### PR TITLE
Disallow removing some metadata fields by remove ingest processor

### PR DESCRIPTION
**Pull request by gaobinlong**
### Description
When removing metadata fields such as `_index`, `_version`, `_id` can cause unexpected result: OpenSearch process crash or `null_pointer_exception`, so it's better we do some check in remove ingest processor to prevent those things from happening. 

In current implementation, the enriched document `IngestDocument` contains these metadata fields: `_index`, `_version`, `_id`, `_version_type`, `_routing`: https://github.com/opensearch-project/OpenSearch/blob/5bd413c588f48589c6fd6c4de4e87550271aecf8/server/src/main/java/org/opensearch/ingest/IngestDocument.java#L83
, in most cases, these metadata fields are used internally, but actually they can be removed by remove ingest processor. Removing these metadata fields are meaningless because these fields have corresponding query parameters in index API, remove ingest processor should not disrupt that API, if users want to remove those fields they should remove the query parameters of index API, remove ingest processor should only focus on the source fields of the incoming documents, this makes the responsibility of index API and remove ingest processor clear.
 
We have two options to prevent removing metadata fields in remove ingest processor, one option is that throwing an `illegal_argument_exception` with an error message like `cannot remove metadata fields...`, another option is to ignore the metadata fields when removing the fields from the source document. But in the long run, we may support `exclude` parameter for the remove processor, so if all the metadata fields don't math the `exclude` pattern, we shouldn't remove the metadata fields, we should ignore them, in order to make this behavior consistently, I prefer the option which ignoring all the metadata fields.

### Related Issues

https://github.com/opensearch-project/OpenSearch/issues/10732

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by gaobinlong** [comment]
Could anyone help to review this PR?

**Comment by reta** [comment]
> Could anyone help to review this PR?

@gaobinlong I am not on board with this change, I believe at least some metadata fields are perfect candidates for removal. Take `_id` field for example: the processor may deliberately remove the `_id` in order to not accept the user supplied ones (for whatever reasons). 

@msfroh do you have an opinion?

**Comment by gaobinlong** [comment]
Removing `_id` and `_routing` maybe useful in some cases, but removing `_index`, `_version` and `_version_type` are meaningless and cause problems, so maybe we can only disallow removing `_index`, `_version` and `_version_type`.

**Comment by reta** [comment]
> Removing `_id` and `_routing` maybe useful in some cases, but removing `_index`, `_version` and `_version_type` are meaningless and cause problems, so maybe we can only disallow removing `_index`, `_version` and `_version_type`.

`_index` probably make sense to disallow, why  removing `_version` and `_version_type` is causing issues?

**Comment by gaobinlong** [comment]
@reta removing `_version` field causes null_pointer_exception, I put some bad cases in the issue:https://github.com/opensearch-project/OpenSearch/issues/10732.

**Comment by reta** [comment]
> @reta removing `_version` field causes null_pointer_exception, I put some bad cases in the issue:#10732.

Got it thanks @gaobinlong , so I think this is what we need to do:
1) Document which metadata fields could and could not be removed  (as per discussions)
2) The Ingest Processor should fail the execution if the fields from above are configured to be removed but could not

Silently ignoring fields is not the best option I believe.

**Comment by gaobinlong** [comment]
> > @reta removing `_version` field causes null_pointer_exception, I put some bad cases in the issue:#10732.
> 
> Got it thanks @gaobinlong , so I think this is what we need to do:
> 
> 1. Document which metadata fields could and could not be removed  (as per discussions)
> 2. The Ingest Processor should fail the execution if the fields from above are configured to be removed but could not
> 
> Silently ignoring fields is not the best option I believe.

Thanks for your suggestion, I have changed the behavior when removing the metadata fields, and also opened an document issue for this change: [#5634](https://github.com/opensearch-project/documentation-website/issues/5634), please help to take a second look at the change, thanks!

**Comment by gaobinlong** [comment]
Flaky test: https://github.com/opensearch-project/OpenSearch/issues/10749

File: modules/ingest-common/src/main/java/org/opensearch/ingest/common/RemoveProcessor.java [hunk 1]
```diff
@@ -79,6 +80,25 @@ public IngestDocument execute(IngestDocument document) {
...
             }
+            // cannot remove _index, _version and _version_type.
+            if (path.equals(IngestDocument.Metadata.INDEX.getFieldName())
+                || path.equals(IngestDocument.Metadata.VERSION.getFieldName())
+                || path.equals(IngestDocument.Metadata.VERSION_TYPE.getFieldName())) {
+                throw new IllegalArgumentException("cannot remove metadata field [" + path + "]");
+            }
+            // removing _id is disallowed when there's an external version specified in the request
+            if (path.equals(IngestDocument.Metadata.ID.getFieldName())
+                && !document.getFieldValue(IngestDocument.Metadata.VERSION_TYPE.getFieldName(), String.class)
```
Comment (reta) [inline modules/ingest-common/src/main/java/org/opensearch/ingest/common/RemoveProcessor.java:91]: `document.getFieldValue(...`) could return `null` (as per java doc):

```suggestion
                && !Objects.equals(document.getFieldValue(IngestDocument.Metadata.VERSION_TYPE.getFieldName(), String.class), VersionType.toString(VersionType.INTERNAL))
```

File: modules/ingest-common/src/main/java/org/opensearch/ingest/common/RemoveProcessor.java [hunk 2]
```diff
@@ -79,6 +80,25 @@ public IngestDocument execute(IngestDocument document) {
...
+                    .equals(VersionType.toString(VersionType.INTERNAL))) {
+                Long version = document.getFieldValue(IngestDocument.Metadata.VERSION.getFieldName(), Long.class);
+                String versionType = document.getFieldValue(IngestDocument.Metadata.VERSION_TYPE.getFieldName(), String.class);
```
Comment (reta) [inline modules/ingest-common/src/main/java/org/opensearch/ingest/common/RemoveProcessor.java:94]: We could probably extract it before `if` statement so we won't be getting the same value twice

**Comment by reta** [comment]
@gaobinlong a few minor comments but LGTM otherwise, thank you!

File: modules/ingest-common/src/main/java/org/opensearch/ingest/common/RemoveProcessor.java (hunk 1 above)
Comment (gaobinlong) [reply modules/ingest-common/src/main/java/org/opensearch/ingest/common/RemoveProcessor.java]: Thanks, I've changed that.

**Comment by gaobinlong** [comment]
Now all gradle checks have passed yet, but the code coverage check is not, that's because of the randomization in our test code.

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@gaobinlong we would need documentation update for this change as well, could you please create the issue at https://github.com/opensearch-project/documentation-website/issues ?

**Comment by reta** [comment]
@gaobinlong and we need manual backport to `2.x` sadly :( please create one

**Comment by gaobinlong** [comment]
Thanks @reta ,I've created a document issue: [5634](https://github.com/opensearch-project/documentation-website/issues/5634) and a backport PR: https://github.com/opensearch-project/OpenSearch/pull/11311.

**Pull request merged**